Intermittent CI failure on textual testbed tests

### Describe the bug

We are seeing intermittent failures in CI on the testbed. The failure appears to be particularly prevalent on the textual-linux backend.

### How to reproduce the bug

Run the testbed test suite using the Textual backend under Ubuntu 24.04 (matching the CI configuration).




### Minimum example code

```shell
briefcase dev -a testbed-textual
```

### Screenshots

_No response_

### Environment details

- Operating system and version: Ubuntu 24.04
- Python version: 3.12
- Software versions:
  - Toga: main as of July 27 2026.



### Logs

[<!-- Paste your log content between the backticks, or upload a file. -->](https://github.com/beeware/toga/actions/runs/29553516152/job/87803265860?pr=4566)

```

=================================== FAILURES ===================================
________________________________ test_parenting ________________________________
Traceback (most recent call last):
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/_pytest/runner.py", line 361, in from_call
    result: TResult | None = func()
                             ^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/_pytest/runner.py", line 250, in <lambda>
    lambda: runtest_hook(item=item, **kwds),
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/_pytest/logging.py", line 865, in pytest_runtest_call
    yield
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/_pytest/capture.py", line 900, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/_pytest/skipping.py", line 268, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/_pytest/runner.py", line 184, in pytest_runtest_call
    item.runtest()
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pytest_asyncio/plugin.py", line 569, in runtest
    super().runtest()
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/_pytest/python.py", line 1707, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_result.py", line 103, in get_result
    raise exc.with_traceback(tb)
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    res = yield
          ^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/_pytest/python.py", line 167, in pytest_pyfunc_call
    result = testfunction(**testargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app_packages/pytest_asyncio/plugin.py", line 905, in inner
    runner.run(coro, context=context)
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app/tests/conftest.py", line 202, in run_until_complete
    return asyncio.run_coroutine_threadsafe(coro, toga.App.app._impl.loop).result()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app/tests/widgets/test_base.py", line 127, in test_parenting
    other_probe.assert_layout(position=(100, 0), size=(100, 200))
  File "/home/runner/work/toga/toga/testbed/build/testbed-textual/ubuntu/noble/testbed-textual-0.0.1/usr/lib/testbed-textual/app/tests_backend/widgets/base.py", line 69, in assert_layout
    assert self.x == self.approx_width(position[0])
AssertionError: assert 0 == 100 ± 5
  
  comparison failed
  Obtained: 0
  Expected: 100 ± 5
---------------------------- Captured stdout setup -----------------------------
Resetting main_window

Constructing Box probe
----------------------------- Captured stdout call -----------------------------
Child should not in layout yet
=========================== short test summary info ============================
FAILED tests/widgets/test_base.py::test_parenting - assert 0 == 100 ± 5
  
  comparison failed
  Obtained: 0
  Expected: 100 ± 5
============ 1 failed, 74 passed, 78 skipped, 29 xfailed in 31.76s =============
```


### Additional context

_No response_

--- a/changes/4568.misc.md
+++ b/changes/4568.misc.md
@@ -1,0 +1,1 @@
+Some failure modes in the textual test probe have been resolved.

--- a/testbed/tests/conftest.py
+++ b/testbed/tests/conftest.py
@@ -199,7 +199,24 @@ class ProxyEventLoop(asyncio.AbstractEventLoop):
             coro = future.coro
         else:
             raise TypeError(f"Future type {type(future)} is not currently supported")
-        return asyncio.run_coroutine_threadsafe(coro, toga.App.app._impl.loop).result()
+
+        # A backend may need to establish a context on the app's event loop
+        # before test code runs. If the probe defines a `test_context`
+        # classmethod, the test is run in that context.
+        backend_module = import_module("tests_backend.app")
+        try:
+            test_context = backend_module.AppProbe.test_context
+
+            async def coro_in_context():
+                with test_context():
+                    return await coro
+
+            test_coro = coro_in_context()
+        except AttributeError:
+            test_coro = coro
+
+        loop = toga.App.app._impl.loop
+        return asyncio.run_coroutine_threadsafe(test_coro, loop).result()
 
     async def shutdown_asyncgens(self):
         # The proxy event loop doesn't need to shut anything down; the

--- a/textual/tests_backend/app.py
+++ b/textual/tests_backend/app.py
@@ -4,6 +4,8 @@ from pathlib import Path
 import pytest
 from textual.app import App as TextualApp
 
+import toga
+
 from .probe import BaseProbe
 
 APP_ID = "org.beeware.toga.testbed-textual"
@@ -20,6 +22,13 @@ class AppProbe(BaseProbe):
     edit_menu_noop_enabled = False
     supports_psutil = True
     beep_delay = 0.1
+
+    @classmethod
+    def test_context(cls):
+        """Establish the context required to safely run test code on the app's event
+        loop.
+        """
+        return toga.App.app._impl.native._context()
 
     def __init__(self, app):
         super().__init__()

--- a/textual/tests_backend/probe.py
+++ b/textual/tests_backend/probe.py
@@ -1,6 +1,8 @@
 import asyncio
 
 import pytest
+from textual._wait import wait_for_idle
+from textual.app import ScreenStackError
 
 import toga
 
@@ -15,22 +17,64 @@ class BaseProbe:
     def approx_height(self, height):
         return pytest.approx(height, abs=self.VERTICAL_SCALE // 2)
 
-    async def wait_for_refresh(self):
-        app = toga.App.app._impl.native
-        loop = asyncio.get_running_loop()
-        refreshed = loop.create_future()
+    async def _wait_for_screen(self, timeout=5.0):
+        """Wait until the screen and all its child widgets have processed the
+        messages currently in their queues.
 
-        if app.call_after_refresh(refreshed.set_result, None):
-            app.refresh(layout=True)
-            await refreshed
+        This duplicates the approach used by Textual's own `Pilot._wait_for_screen`.
+        """
+        native = self.app._impl.native
+        try:
+            screen = native.screen
+        except ScreenStackError:
+            return
+
+        children = [native, *screen.walk_children(with_self=True)]
+        count = 0
+        count_zero_event = asyncio.Event()
+
+        def decrement_counter():
+            """Decrement internal counter, and set an event if it reaches zero."""
+            nonlocal count
+            count -= 1
+            if count == 0:
+                # When count is zero, all messages queued at the start of the
+                # method have been processed.
+                count_zero_event.set()
+
+        for child in children:
+            if child.call_later(decrement_counter):
+                count += 1
+
+        if count:
+            # Wait for the count to return to zero, or a timeout, or an exception
+            wait_for = [
+                asyncio.ensure_future(count_zero_event.wait()),
+                asyncio.ensure_future(native._exception_event.wait()),
+            ]
+            _, pending = await asyncio.wait(
+                wait_for,
+                timeout=timeout,
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+            for task in pending:
+                task.cancel()
+
+            timed_out = len(wait_for) == len(pending)
+            if timed_out:
+                pytest.fail(
+                    "Timed out while waiting for widgets to process pending messages."
+                )
 
     async def redraw(self, message=None, delay=0, wait_for=None):
+        # This implementation is strongly inspired by Textual's Pilot.pause(),
+        # but adding `wait_for` handling.
         if toga.App.app.run_slow or wait_for:
             delay = max(1, delay)
 
-        await asyncio.sleep(0)
-
         print("Waiting for redraw" if message is None else message)
+        await self._wait_for_screen()
+        await wait_for_idle(0)
         if delay or wait_for:
             if toga.App.app.run_slow or wait_for is None:
                 await asyncio.sleep(delay)
@@ -41,7 +85,7 @@ class BaseProbe:
                     await asyncio.sleep(delta)
                     interval += delta
 
-        await self.wait_for_refresh()
+        self.app._impl.native.screen._on_timer_update()
 
     def assert_image_size(self, image_size, size, screen, window=None):
         pytest.skip("Image size assertions are not implemented on Textual.")

--- a/textual/tests_backend/probe.py
+++ b/textual/tests_backend/probe.py
@@ -2,7 +2,6 @@ import asyncio
 
 import pytest
 from textual._wait import wait_for_idle
-from textual.app import ScreenStackError
 
 import toga
 
@@ -26,8 +25,8 @@ class BaseProbe:
         native = self.app._impl.native
         try:
             screen = native.screen
-        except ScreenStackError:
-            return
+        except Exception:
+            return False
 
         children = [native, *screen.walk_children(with_self=True)]
         count = 0
@@ -65,6 +64,13 @@ class BaseProbe:
                 pytest.fail(
                     "Timed out while waiting for widgets to process pending messages."
                 )
+
+            # We've either timed out, encountered an exception, or we've finished
+            # decrementing all the counters (all events processed in children).
+            if count > 0:
+                return False
+
+        return True
 
     async def redraw(self, message=None, delay=0, wait_for=None):
         # This implementation is strongly inspired by Textual's Pilot.pause(),

